Entity set aggregation

athoscouto opened the pull request:
### Issues
*This pull request adds infrastructure to support the aggregation of entity sets properties, as described in [#893](https://github.com/OData/WebApi/issues/893) of WebAPI project.*

### Problems
1. *We need to try properties of aggregated entity sets as single value nodes. The problem is that since entity sets are collections, they cannot be parents of a single value node.*
2. *There's an ambiguity between the call of a function and the aggregation of a entity set property. As per the OData protocol, function and action names should be prefixed with the namespace, so there should be no ambiguity. However, ODL is not fully compliant with the protocol, so there's ambiguity while making the distinction*

### Checklist (Uncheck if it is not completed)
- [x] *Test cases added.*
- [x] *Build and test with one-click build and test script (mostly) passed.*
- [x] *Check if approach to enhance the parser was appropriate.*
- [ ] *Check if approach to extend syntactic tree was appropriate and how to modify it in order to not interfere with the restriction that single value nodes have to have single value parents (that is justifiable in most of the codebase).*
- [x] Add support to access inner properties.

### Additional work necessary
*There's still one problem to address. I had to modify the syntactic parser to allow SingleValueNodes to have parents that aren't a single value. This is needed because when we aggregate a property from a entity set it is treated as a single value, even tough it is derived from a collection. 
This change break tests ShouldThrowIfPropertyAccessSourceNotSingleValueIfRequiredOnBinaryOperand and ShouldThrowIfPropertyAccessSourceNotSingleValueIfUsedOnOrderBy from OData.Full solution. 
In OData.Lite solution the tests **InnerPathTokenBinderShouldFailIfPropertySourceIsNotASingleValue** and **WhenAnyIsLedBySingletonNavigationPropertyShouldFail** were affected. 
We should decide if **AggregatedCollectionPropertyNode** should only be allowed in aggregate clauses or if it's better to relax the tests.*

GeoffCox commented on src/Microsoft.OData.Core/Build.Net35/Microsoft/OData/Core/UriParser/Aggregation/BaseTokens.cs:
```diff
@@ -18,4 +18,11 @@ namespace Microsoft.OData.UriParser.Aggregation
     internal abstract class ApplyTransformationToken : QueryToken
```
Even though these classes are super tiny, I prefer 1 file per class.  It becomes difficult to tell if a class should be in BaseTokens.cs based on if it is a root base class or an intermediate base class.

athoscouto replied:
Done.

GeoffCox commented on src/Microsoft.OData.Core/UriParser/Aggregation/AggregateExpression.cs:
```diff
@@ -8,43 +8,63 @@ namespace Microsoft.OData.UriParser.Aggregation
     /// A aggregate expression representing a aggregation transformation.
     /// </summary>
-    public sealed class AggregateExpression
+    public abstract class AggregateExpression
```
Abstract classes usually have an abstract method for the derived class to implement.  Consider a protected default constructor if you don't want callers instantiating this class.

athoscouto replied:
I'm not sure I'm getting this comment. The only constructor of the class is protected. Isn't the default constructor present only in the absence of other constructors?

GeoffCox commented on src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs:
```diff
@@ -447,6 +453,24 @@ internal string AdvanceThroughBalancedParentheticalExpression()
             return expressionText;
         }
 
+        internal void TakeSnapshot()
```
Nit: It would be good to have TextPosition in the method names here to make it clear what is being snapshotted.

Remind me to show you a way to implement an IDisposable return type to allow callers to nest snapshots.

athoscouto replied:
Done.

GeoffCox commented on test/FunctionalTests/Tests/DataOData/Tests/OData.Common.Tests/PublicApi/PublicApi.bsl:
```diff
@@ -6148,6 +6163,12 @@ public sealed class Microsoft.OData.UriParser.WildcardSelectItem : Microsoft.ODa
+
 public enum Microsoft.OData.UriParser.Aggregation.AggregationMethod : int {
 	Average = 3
 	CountDistinct = 4
```
What happened to 2?

athoscouto replied:
Nothing. GitHub just skipped some lines because there was no changes in it.
If you look at [here](https://github.com/athoscouto/odata.net/blob/entity_set_aggregation/test/FunctionalTests/Tests/DataOData/Tests/OData.Common.Tests/PublicApi/PublicApi.bsl#L6175), you'll see that number this enum is unchanged. This number 1 showing is from *enum Microsoft.OData.UriParser.Aggregation.TransformationNodeKind*, while 3 and 4 here belong to *enum Microsoft.OData.UriParser.Aggregation.AggregationMethod*.

kosinsky committed:
Support dynamic properties in $apply

mikepizzo committed:
Support Reading/Writing Optional Parameters

mikepizzo committed:
Support Reading/Writing Optional Parameters
Support selecting best function overload based on optional parameters

mikepizzo committed:
Update tests, validation.
Support annotating ReturnType (and ignore unexpected annotations elsewhere)

robward-ms committed:
Add script to cleanup databases for automated builds

robward-ms committed:
Fix style cop issue from #808.

AlanWong-MS committed:
Updating build script for Windows Store test solution

Updating build.ps1 to skip building the WindowsStore test solution if VS2013 is not installed. This is due to the fact that Windows 8.0 apps are not supported with VS2015.

Also updating the solution file to build the projects in order in the event of race conditions during build.

And adding .NET Core packages to nuget for the .NET Standard solution.

AlanWong-MS committed:
Fix indentation in build.ps1

mikepizzo committed:
Fix Issue #814

biaol-odata committed:
Updating readme for adopting standard Git branching scheme

robward-ms committed:
Merge branch 'odata-master' of https://github.com/biaol-odata/odata.net

biaol-odata committed:
Merge remote-tracking branch 'origin/ODataV4-7.x'

xuzhg committed:
Update the standard link

mtroth-microsoft committed:
Implementation of $compute

mikepizzo committed:
Added files to legacy projects and fixing CA issues.

mikepizzo committed:
Added files to legacy projects and fixing CA issues.

mikepizzo committed:
Added files to legacy projects and fixing CA issues.

mikepizzo committed:
Added files to legacy projects and fixing CA issues.

mikepizzo committed:
Merge pull request #1 from OData/OptionalParameters

Added files to legacy projects and fixing CA issues.

mikepizzo committed:
Merge branch 'ODataV4-7.x' of https://github.com/OData/odata.net into ODataV4-7.x

mikepizzo committed:
Merge branch 'OptionalParameters' of https://github.com/MikePizzo/odata.net into ODataV4-7.x

# Conflicts:
#	src/Microsoft.OData.Edm/Csdl/Parsing/Common/XmlElementParser.cs

mikepizzo committed:
Fix test issues

robward-ms committed:
Update the expired certificate for the WindowsStore test project.

mikepizzo committed:
Fix tests and CA issues

mikepizzo committed:
Fix legacy project files

mikepizzo committed:
Fix legacy project files

mikepizzo committed:
Fix stupid project issues.

mikepizzo committed:
Fix stupid project issues.

mikepizzo committed:
Fix stupid project issues.

mikepizzo committed:
Fix stupid project issues.

mikepizzo committed:
Merge branch 'ODataV4-7.x' of https://github.com/OData/odata.net into ODataV4-7.x

mikepizzo committed:
Fixing so build works on .Net35.

mikepizzo committed:
Update JsonLightInstanceAnnotationWriter.cs

Fix spacing around if statement for CA.

kosinsky committed:
Merge PR #816, issue #760: Support average for dynamic properties
Dynamic properties don't have type.

mikepizzo committed:
Initial commit of untyped support.
-Supports reading untyped structural properties as dynamic complex types.
-Supports reading untyped primitive properties as bool, decimal, or string.
-Preserves reading/writing type information

robward-ms committed:
Cleanup skipped tests in Microsoft.OData.Core.Tests.

Skipped tests have been resolved by:
1.) Unskipping the test
2.) Modifying the test to match current product behavior
3.) Commenting out the test where matching product work is not complete, add the matching GitHiub issue
4.) Removing the test

Commit:
Merge PR #830: Add Issue Triage Guidance

mikepizzo committed:
Reading/Writing Untyped as structure

mtroth-microsoft committed:
Add Support for PrimitiveType casts

biaol-odata committed:
Remove unreachable code block for collection reader.

biaol-odata committed:
Update test coverage for ODataCollectionReader.

Commit:
Merge PR #829: Update the markdown

robward-ms committed:
Fix Code Analysis issue from PR #844.

biaol-odata committed:
Restore setter method in PR #844

xuzhg committed:
Add test cases for the ODataAtomErrorDeserializer coverage, PR #845

xuzhg committed:
modify the test case to make VSTS build happy

AlanWong-MS committed:
Re-enable priority test cases in Microsoft.Data.Web.UnitTests suite (#842)

* Modifying and enabling test cases in Microsoft.Data.Web.UnitTests. These test cases cover scenarios that we've deemed higher priority. A separate commit will address the rest of the skipped tests.

xuzhg committed:
Clean up codes related to Spatial type deseializer

xuzhg committed:
Add odata.core.test project into full solution

xuzhg committed:
Add the xunit.runner.visualstudio package

xuzhg committed:
Raw Value serializer output json string for Spatial type, PR #856

xuzhg committed:
Clean codes in EdmLibExtensions of OData.Core

Squashed commit of the following:

commit 73af3af95c7044668f3032584e0bc0a790cb9db5
Author: Sam Xu <saxu@microsoft.com>
Date:   Tue Jun 27 10:28:57 2017 -0700

    Remove unused comments

commit 5eb690953598c7d17c91fd9af61df96616b8c0b1
Author: Sam Xu <saxu@microsoft.com>
Date:   Thu Jun 22 22:03:16 2017 -0700

    Clean codes in EdmLibExtensions for coverage

0xced committed:
Assert DateTimeOffset equality on DateTimeOffset object

This will make the test pass even on systems configured with a non-US culture

kosinsky committed:
Adding support to entity set aggregations.

kosinsky committed:
Adding support to entity set aggregations.

mtroth-microsoft committed:
Fixing Build Cop warning

xuzhg committed:
Clean the skipped test cases, PR #865

biaol-odata committed:
Increase coverage for JsonLight in Microsoft.OData.Core (#862)

AlanWong-MS committed:
Improve test coverage and fix bugs for OData.Core UriParser feature (#863)

- Adding new tests to fill priority test gaps
- Fixing bugs found
- Removing src/Microsoft.OData.Core/UriParser/Parsers/OpenTypeMethods.cs as dead code; there is a similar class OpenTypeMethods defined and used in the test code

robward-ms committed:
Standardize on EntityFramework 5.0 for tests to reduce race conditions
for EntityFramework dependent tests.

robward-ms committed:
Modify database strategy to eliminate Attach. This approach takes longer to run
but eliminates one of the reasons tests must be run ad Administrator.

See this for details:
https://stackoverflow.com/questions/2330439/access-is-denied-when-attaching-a-database

robward-ms committed:
By default, make build.cmd run all tests as to ensure nothing is missed.
Optionally, run build.cmd quick to run only the unit tests.
Comment out Asserts to allow tests to pass in Debug configuration.
Add TestStartup.cmd to prepare for test runs in the build environment.
Remove TestCategoryAttributeCheck.csproj from the solution as it provides little value.
Rename a misspelled test.

biaol-odata committed:
Cleanup CodeAnalysis exclusions for ODL.Core (#884)

*ODL.Core: Cleanup top level globalSuppressions config.

*ODL.Core: Turning off all AC suppressions.

*ODL.Core: Turning off CA3053.

* ODL.Spatial and ODL.Edm: CA exclusions analysis of suppressed security warnings.

xuzhg committed:
Comment out the skipped tests, detail see PR #882

mtroth-microsoft committed:
Fix FxCop issue in build

biaol-odata committed:
ODL-Spatial, ODL-Edm, ODL-Client: cleanup obsolete AC warnings suppressions. (#891)

biaol-odata committed:
CA13xx(Globalization) warning message cleanup for Odata.net (#892)

* CA13xx(Globalization) warning message cleanup for Odata.net

* CR updates.

mikepizzo committed:
Merge pull request #828 from mikepizzo/OptionalParameters

Optional parameters

xuzhg committed:
Cleanup Microsoft.Reliability in OData.Core, OData.Client, Detail see PR

biaol-odata committed:
CA exclusion analysis: MSInternal warnings: removing CA908. (#893)

xuzhg committed:
Clean the extra skipped test cases from VSTS, PR #897

robward-ms committed:
Update NuSpec files for generating Nuget packages.
 - Remove resource dlls from NuGet packages.
 - Add $ProductRoot$ variable to specify product location.
 - Update the license link to be consistent.
 - Update the release note Url to include v7.
 - Add GetNugetPackageMetadata.proj to set build variables to pass to Nuget.exe pack command.
 - Remove *.nuProj files; package using *.nuspec files directly.
 - Remove *Debug.NuSpec; only *Release.nuspec and *Nightly.Release.nuspec are needed.

biaol-odata committed:
CA exclusions: CA18xx category for ODL (#900)

* CAExclusions: cleanup CA18xx (Performance).

* CA 18xx category: EDM

* CA 18xx catetory: ODL-Client.

* Picking up couple StyleCop issues in current build.

* StyleCop fixes.

biaol-odata committed:
Add build badge to show latest build status. (#906)

robward-ms committed:
Update Versioning.props
     - Add date code as the revision field of the binaries
     - Re-organize the file for easier maintenance

mikepizzo committed:
-Improved support for untyped in contextUri
-Added more tests
-Improved error codes
-Addressed code review issues

biaol-odata committed:
Add nightly build badge (#911)

* Add nightly build badge

* CR comment: change table to markdown.

onlyann committed:
change EdmStructuredSchemaElements to a ConcurrentDictionary (#905)

* change EdmStructuredSchemaElements to a ConcurrentDictionary
* change default property value to be VS2013 syntax compliant
* add ConcurrentDictionary shim to support Windows Phone 8.0

AlanWong-MS committed:
Re-organizing solution file layout for ODL master branch (#907)

Re-organizing solution file layout for ODL master branch
- Renaming solutions from Microsoft.(Test.)OData.* to OData.Product/Tests to reduce name size
- Creating new light solutions for each supported framework (Net35, Net45, and NetStandard) and corresponding unit tests.
- Reducing number of solutions by moving most tests to Microsoft.OData.Full.sln, which has been renamed to OData.Tests.E2E.sln. This solution builds on Net45
- There are two .NetStandard solutions. The product one contains only the product, and the test one contains both the product and unit tests. The product-only solution will eventually be removed once ODL migrates to Visual Studio 2017; this is a current workaround to pre-released version of .NET Core tooling in VS2015.
- Fixing bug in test where dashes in a hostname causes SQL statements to not execute properly. Replacing any '-' (there shouldn't be any other special characters) found in hostnames with '_'.
- Updating docs to reflect solution changes

robward-ms committed:
Start 7.4.0 beta

lensgrinder committed:
Add Compute to ParseUri.

xuzhg committed:
Change the performance script to make local build

xuzhg committed:
add a script to run the performance on VSTS

robward-ms committed:
Update nightly feed details.

xuzhg committed:
modify the performance scripts

xuzhg committed:
Modify the official lib path and add performance script

xuzhg committed:
modify the throw message of regression

xuzhg committed:
modify the project file to let VSTS build happy

xuzhg committed:
modify performance script typo

biaol-odata committed:
Policheck run on VSTS for ODL-master (#919)

* Policheck fixes

* Add the cmd to run policheck

* remove the RunPoliCheck.cmd

* Adding the exclusion config for termId 211972

* create a Policheck subdir for each branch.

* Add script to run policheck.

* move policheck exclusion config out of tools (to images dir)

AlanWong-MS committed:
Migrating .NET Core tests and .NET Standard source projects to VS2017 (#921)

- OData.Tests.NetStandard.sln is now OData.Tests.NetStandard.VS2017.sln to indicate that it is to be opened with VS2017. Because we have a massive number of projects to migrate from VS2015, we will do them incrementally and use the VS2017 suffix to indicate which solutions are VS2017. Once we are done with the migration, we'll remove the suffix.
- Replaced .xproj with new VS2017 .csproj for .NET Core tests. You will need to have VS2017 installed if you wish to run the unit tests in .NET Core. As a result of the migration, also cleaning up the project.jsons and related files.
- Added VS2017 .csproj versions of the .NET Standard product source. You may still launch OData.NetStandard.sln to modify and build .NET Standard with VS2015. However, once we are in a state where we can flip the switch to be 100% VS2017, OData.NetStandard.sln will also be in VS2017.
- Deleted src\Microsoft.OData.Core\PortabilityExtensionMethods.cs as it is not used
- Disabling a scenario for the SecurityPartialTrustTest case as it is inconsistently passing.
- Fixing a few stylecop errors
- Updating build.ps1 to build and run the .NET Core tests if VS2017 and dotnet CLI are installed
- Updating the logic in the test results calculation in build.ps1. It was previously parsing each line in the test logs to count number of passed/skipped tests. The dotnet CLI doesn't provide this output, so the script would missing the passed/skipped tests for .NET Core tests. Modifying the algorithm to parse the "Total tests..." line for total passed/skipped tests, which is generalized across both VSTest and dotnet.

xuzhg committed:
Change the performance official project reference

xuzhg committed:
Add a performance running scripts

xuzhg committed:
modify the performance test to output GC allocation

AlanWong-MS committed:
Removing tests for Windows 8.0 apps (#932)

Removing Microsoft.Test.OData.Tests.Client.Portable.WindowsStore.csproj and OData.Tests.WindowsStore.VS2013.sln due to complexity of maintaining VS2013. With this removal, the VS2013 will no longer be an installation requirement for comprehensive testing.

Per .NET Standard support matrix, .NET Standard 1.1 supports Windows 8.0 apps but we are making a conscious decision to remove the tests for that platform. This allows us to simplify the number of installations needed to develop for OData and achieve our overall goal of migration to the latest version of Visual Studio (2017 as of writing). Further, Windows 8.0 has been long superseded by Windows 8.1 (and 10), so there is low risk in removing such tests. Should any developer wish to run these tests, he/she may review the commit history to retrieve the test project.

robward-ms committed:
Remove passwords from source and test code.

greghug committed:
Avoid parsing inside a lock in CustomUriLiteralParsers

biaol-odata committed:
remove IIS Local server test cases due to timing issue (#935)

biaol-odata committed:
Disable additional test case data subjected to IIS local server instance racing (#946)

* Disable additional test case data subjected to IIS local server instance racing in build machine for code coverage debug run.
- SecurityPartialTrustTest
- HttpContextServiceHostTunnelingTest
- RequestUriCaseInsensitive

* CR comment

xuzhg committed:
Issue#949,PR#954:ODataUriExtensions.BuildUri ignores $skiptoken and $deltatoken

0xced committed:
Make DataServiceQuerySingle.GetValueAsync consistent with GetValue

Use SingleOrDefault() instead of Single() to get the same behavior as GetValue.

Fixes #698

biaol-odata committed:
porting to master branch: multipart/mixed refactoring

biaol-odata committed:
Address some CR comments

bhaarathms committed:
Add support for ENUM keys

biaol-odata committed:
fix the CustomUriLiteralPrefixedTests to handle interference to static prefix registration from other test runs

AlanWong-MS committed:
Update PULL_REQUEST_TEMPLATE.md for markdown

- Updating PULL_REQUEST_TEMPLATE.md for markdown support in GitHub

bhaarathms committed:
Use IsEnum() method

biaol-odata committed:
Address addtional CR comments: cleanups, adjusting some APIs between base and derived classes; Remaining CR items are related to ongoing refactoring discussion

bhaarathms committed:
Add tests

bhaarathms committed:
Fix typo

biaol-odata committed:
Address refactoring comments: moving some utility APIs between base and derived class.

biaol-odata committed:
refactoring: ReadImplementation & associated APIs (for multipart batch)

Commit:
Escaped special meaning characters remain escaped when parsing the context uri +UT

manivannanse committed:
Merge pull request #961 from bhaarathms/SupportEnumKeys

Add support for ENUM keys

manivannanse committed:
Merge pull request #969 from OfirArviv/master

Escaped special meaning characters remain escaped when parsing the co…

biaol-odata committed:
Updates for CR comments: removing IsInsideSubBatch, CanCreateOperationRequestMessageVerifierCommon and StartBatchOperationContent

biaol-odata committed:
refactoring: base batch writer: removed protected CreateOperationRequestUriWrapper, AddToPayloadUriConverter, VerifyCanCreateOperationRequestMessageImplementation; changed to private RememberContentIdHeader; addressing CR comments & cleanups.

robward-ms committed:
Update rolling build badge to use new build definition id.

biaol-odata committed:
MIME batch reader base class refactoring: merging changes from mikep's https://github.com/biaol-odata/odata.net/pull/1/commits/01049e263e894f54db8f4765a1a007d5488e41f3 to tip of PR. ...WIP

biaol-odata committed:
increment batch size, add contentId on next read; etc

biaol-odata committed:
Multipart/mixed refactoring: moved logic around a bit more; tighten logic difference btw <=V4 and >V4.

AlanWong-MS committed:
OData Client NetStandard test coverage (#971)

### Issues
This pull request addresses #731 and introduces NetCore test cases to cover the new NetStandard OData Client library for ODL 7.x.

### Description
The following projects are built on NetCore from their NetFx versions and they test the OData Client 7.x NetStandard library:
- Microsoft.OData.Client.Tests.Netcore
- Microsoft.OData.Client.TDDUnitTests.NetCore
- Microsoft.Test.OData.Tests.Client.NetCore

The estimated code coverage percentage is 74% (estimated because Visual Studio 2017 cannot natively obtain the numbers for NetCore tests yet as of version 15.3.5).

Other changes include:
- Removal of Microsoft.Test.OData.Framework.NetStandard.csproj as it's no longer needed.
- Microsoft.Test.OData.Tests.Client.NetCore utilizes an out-proc test service and communicates with it via a service wrapper app, aptly called ServiceWrapperApp.exe. This change was due to the fact that the test services had too many projects to port to NetCore, coupled with the fact that porting NetFx to NetCore takes a tremendous effort. Thus the test project Microsoft.Test.OData.Tests.Client is separated into a test client (Microsoft.Test.OData.Tests.Client.NetCore.csproj) in NetCore and a test service wrapper (ServiceWrapperApp.csproj) that launches the test service written in NetFx. The two projects communicate with each other through a new service interface.
- Many of the existing OData Client test cases call synchronous APIs, which are not applicable to the portable/NetStandard version of the library, and therefore they are excluded from the NetCore test cases.
- ReadMe updated to address changes

mikepizzo committed:
Demonstrates clean factoring of multipart/mixed out of non-multipart/mixed data types: (#2)

refactoring related to batchBoundary in ODataMessageReader/Writer.

robward-ms committed:
Added IgnoreClientProperty attribute in order to ignore client-side entity properties contained in partial classes

biaol-odata committed:
manual merge from https://github.com/biaol-odata/odata.net/pull/3 with modifications making it less hacky

AlanWong-MS committed:
Updating Nuget package for OData Client to publish only NetStandard binaries (#981)

- Updating Nuget package for OData Client to remove NetFx45 and PortableLib binaries
- Fixing bug in ServiceWrapperApp where exception is ignored

biaol-odata committed:
Moved input/output context for multipart/mixed batch to separat files; moved ODataBatchFormat and renamed to ODataMultipartMixedBatchFormat.

robward-ms committed:
Add V7 suffix to Net35 binaries.

robward-ms committed:
Update to 7.5.0

robward-ms committed:
Update .gitignore with new FX35 xml file names.

bhaarathms committed:
Support enum to string comparision (#980)

Support enum to string comparision

biaol-odata committed:
Adding Json Batch support in ODL-Core with additional E2E tests (6.x tests porting WIP).

biaol-odata committed:
Porting additional tests from 6.x to master. Also include new Version401 and error->string conversion functions needed by ODL-client in future plan.

biaol-odata committed:
minor cleanups in classes related to stream

mikepizzo committed:
Demonstrates removing extra complexity around body content reader stream.

biaol-odata committed:
Copyright fixes for nuspec.

biaol-odata committed:
Updating CopyRight info in nuspec files -- additional 2 files

tchuan committed:
Use ConcurrentDictionary in all platforms to make client edm model thread safe (#995)

- Use ConcurrentDictionary in all platforms, except Windows Phone 8.0 (ConcurrentDictionary not supported), to make Client Edm model thread safe.

mikepizzo committed:
Add support for 4.01 Delta Format (#989)

* Initial Commit of writing new delta format
-adds support for 4.01 version
   -defaults to not writing odata prefix
-support for writing 4.01 delta format
   -nested delta feeds
   -for deleted entries
      -@removed: {reason: xyz}
      -deleted entry can include properties
      -@id, optional if keys are included in payload

* Fixes for contexturl, tests

* Code review cleanup

* Merge cleanup

* Implement reading deltas in ODataReader.

* Address Comments from CodeReview.

biaol-odata committed:
Double-checking implementation alignment with OData v401 protocol:
- Support for base64url encoded binary content for operations;
- Support for textual content for operations;
- Lower-case header names in operation message headers;
- Batch writer dependsOn attribute for request;
- Correlation of request id and atomic group id in request and response;
- Req id uniqueness scope difference btw V4 and V401 for multipart(as a flag, add failing test BatchJsonLightContentIdUniquenessScopeForMultipartV401ShouldPass), v401 for Json Batch;
- Disallow data request inside changeset for both V4 and V401 independent of formats;

biaol-odata committed:
Updates for dependsOnIds public interface to be thru create...Message method.

biaol-odata committed:
Enable test BatchJsonLightContentIdUniquenessScopeForMultipartV401ShouldThrow that matches the OData v4.01 protocol.

biaol-odata committed:
non-functional cleanups -- comments, spelling checks, namespaces.

biaol-odata committed:
Adding test for different orders of headers and body stream content

AlanWong-MS committed:
Adding test cases for multipart/mixed and JSON batch scenarios (#5)

* Adding test cases for multipart/mixed and JSON batch scenarios for request write/read

* Updating per code review

biaol-odata committed:
Addressing CR comments -- WIP.

biaol-odata committed:
Addressing all CR comments + introducing dependsOnIds to multipart/mixed batch with backward compatibility.

mikepizzo committed:
Fix property validation issues (#1010)

biaol-odata committed:
Manual merge w/ resolved conflict for additional test cases from Alan's PR https://github.com/biaol-odata/odata.net/pull/6.

biaol-odata committed:
Add stress tests exercising 40GB data stream * round trip -- need to enable manually to run since tests run takes 1+ hr.

biaol-odata committed:
Add perf tests exercising Json & Multipart batch formats with large payload.

biaol-odata committed:
Resolved most latest CR comments

exposing the contentId and groupId for MultipartMixed batch's operation response message.
still need to work on dependsOnIds for MultipartMixed.

biaol-odata committed:
rollback previous impl for dependsonIdsForMultipart.

biaol-odata committed:
API prolog updates.

biaol-odata committed:
Updating test's error message to align with previous API prolog chagnes.

biaol-odata committed:
Update publicAPI to align with API prolog changes.

biaol-odata committed:
FxCopFixes: use ToLowerInvariant.

robward-ms committed:
Update rolling build badge

AlanWong-MS committed:
Adding OData Client Net45 binary (#1018)

* Adding back the OData Client binary for Net45 framework per feedback from community.

biaol-odata committed:
dependsOn Ids for Multipart/Mixed Batch

- Reader extracts contentId header value for both changeset and top-level requests;
- Uses DependsOnIdsTracker to track request ids in Multipart/Mixed batch;
- DependsOnIds validation is performed only for JsonBatch since it is client provided; not for Multipart/Mixed since it is not provided by client.

xuzhg committed:
Fix the issue when calculate the context uri with complex type property access after operation

xuzhg committed:
Change the context uri for operation and operation import and add more test cases to verify

xuzhg committed:
resolve the comments and add extra test to verify the context uri builder

biaol-odata committed:
CR comments.

xuzhg committed:
resolve the comments

xuzhg committed:
modify the error text file for TT generator

mikepizzo committed:
Fix Issue 1193; add back extension function to derived class. (#1025)

* Fix Issue 1193; add back extension function to derived class.

* Update public api baseline

biaol-odata committed:
Addressing CR comment: using null dependsOnIds of MultipartMixed batch as a flag to uniquely indicate implicit dependsOnIds case

xuzhg committed:
Add the enum member into the expression check and don't return the ExpressionNotValidForTheAssertedType error message

xuzhg committed:
resolve the comments and modify the public api test

xuzhg committed:
modify the Public Api Test

xuzhg committed:
modify the failed test for the enummember expression validation

biaol-odata committed:
Addressed addtional CR comments, primarily:
- Changing to use payloadUrlConverter to validate referencing uri for the case of Multipart batch with implicit dependsOnIds.
- Tests added.

biaol-odata committed:
Restricting BatchOperationRequestMessage.dependsOnIds from List<string> to IEnumerable<string>, not allowing element updates.

robward-ms committed:
Update LICENSE.txt

mikepizzo committed:
401 format fixes (#1038)

* Fix Issue 1193; add back extension function to derived class.

* Update public api baseline

* Address remaining OData 4.01 payload requirements

* Improve handling of odata prefix in odata metadata and streaming type parameters.

biaol-odata committed:
Update changeset id to use whole boundary value.

kosinsky committed:
Merge with Rob's to fix StyleCop

kosinsky committed:
Merge remote-tracking branch 'full/master' into entity_set_aggregation

Commit:
Issue #1045 add support DataUriExtensions.BuildUri $apply

xuzhg committed:
Add the build-in abstract type into Edm core model.
1. Edm.ComplexType
2. Edm.EntityType
3. Edm.PrimitiveType
4. Edm.*Path, include Edm.AnnotationPath, Edm.PropertyPath,
Edm.NavigationPropertyPath

xuzhg committed:
Remove the NavigationPropertyEntityMustNotIndirectlyContainItself rule

xuzhg committed:
add the EdmPathTypeReference.cs into Net35 and NetStandard project

xuzhg committed:
Make the e2e test cases pass

xuzhg committed:
Modify test case after removing the rule from the set

xuzhg committed:
resolve the comments

biaol-odata committed:
Small test update and add getter&Setter for EnableNoDollarSignPrefixSystemQueryOption in
* ODataUriResolver
* ODataUriParserConfiguration

xuzhg committed:
Add KeyAsSegmentSupported annotation term to Capabiliites vocabulary

xuzhg committed:
Resolve the typo errors

AlanWong-MS committed:
Adding new OData Client Code Gen for VS2017 (#987)

- Adding VSIX files for OData Code Generator VS extension. Migrating OData.CodeGen.sln to VS2017
- Adding new VSIX project for OData Client Code Gen for VS2017
- Reworking the CodeGen solution to use new VSIX project format. Migrating legacy tests to the E2E solution. Removing a Windows Store test project as we're publishing to the Visual Studio Marketplace and not Windows Store now
- Reducing the size of the VSIX package from ~15mb to 1mb by referencing release binaries as opposed to debug ones and removing embedded NuGet packages. The solution references the ODL projects, allowing the VSIX project to reference the latest versions of ODL during build time. The ODL release builds are disabled within OData.CodeGen.sln because the build process of the solution was overwriting the ODL binaries prepared from a previous build process. Other attempts included using the build flag /p:BuildProjectReferences=false and specifically building the csproj, but either they were ineffective or required too many changes. A caveat to this approach is that OData.Net45.sln must be built first before building the OData.CodeGen.sln.
- Updating README.md and build.ps1 to reflect changes
- Updating test cases to remove hardcoded version numbers
- Updating ItemTemplates to be dummy files.

xuzhg committed:
Resolve the comments

xuzhg committed:
Add a comments in the CapabiliitesVocabularyModel.cs

xuzhg committed:
remove the edmcoremodelconstants.cs and modify the Core Model element constructor

xuzhg committed:
Add the validation rules to the abstract types

xuzhg committed:
Fix the public api failing test

xuzhg committed:
Resolve the typo and add the validation rule for enum type

xuzhg committed:
add the validation rule for path type property in declaring type of entity set, singleton or type of navigaiton property

xuzhg committed:
Add the test cases to verify the validation rule works for recursive type with path type property

Commit:
Issue #1045 add support $apply compute expressions

Commit:
Issue #1045 fix solution file visual studio version

xuzhg committed:
Modify the duplicated annotation validation issue

xuzhg committed:
Add the Authoritation annotation into core edm model

xuzhg committed:
resolve the typo in GlobalSuppressions.cs

xuzhg committed:
Remove unused test codes

robward-ms committed:
Update AuthorizationVocabularies.xml

Update Authorizations to apply to only EntityContainer.
Update Authorization to list name then description.
Add SecuritySchemes term.

mikepizzo committed:
Support annotations in $Select (#1063)

* Support annotations in $Select

* Fix case insensitivity:
1) If exact match, use it
2) When doing case insensitive, search across referenced models
3) Don't reset case insensitivity on Parser Initialize.

Added test cases for AnnotationSegment and case insensitivity.
Fixed test cases so that they didn't rely on ordering for correct case insensitivity setting.

* Minor fixes from CR

Commit:
Issue #1045 fix build error OData.Net35.sln

voronov-maxim committed:
Issue #1045 fix build error OData.Net45.sln

mikepizzo committed:
Add support for AnnotationSegment to PathSegmentHandler. (#1075)

Commit:
Odata v4 7.x - Fix #530 LINQ query generation with Date functions produces weird urls (#1068)

* Fix #530 LINQ query generation with Date functions produces weird urls

* Fix #530 LINQ query generation with Date functions produces weird urls - More tests

robward-ms committed:
This change adds nuget package testing. (#1080)

This intention is to provide a small test project that can ingest the latest NuPkg's
from the build, restore them, build the project which results in a testable Dll. These
tests are very limited in scope, i.e. they only test the Dlls ability to load and
perform basic functions; the functional tests will test the full functionality.

Commit:
Issue #1045 fix test Microsoft.Test.Taupo.OData.Common.Tests.ODataPublicApiTests.PublicApiTest

robward-ms committed:
Fix Policheck issue for package tests. (#1083)

biaol-odata committed:
ExpressionLexer fix for parameter alias token in dotted expression

AlanWong-MS committed:
Updating package ID to match the existing VSIX ID in the marketplace (#1084)

Updating package ID to match the existing VSIX ID in the marketplace

biaol-odata committed:
Removed checking of annotation prefix; Added test with padding spaces

robward-ms committed:
Enable updating top-level properties to null.

Support for writing top-level properties with null values at
the time @odata.null:true was removed. The 4.0 spec states in
1.2.3 Requesting Individual Properties that if the property is
single-valued and has the null value, the service responds with
204 No Content. The library is correct in throwing an exception
when asked to write a null top-level values.

This change adds support for the scenario in the following ways:

1.) Writing a null top-level value still throws but with a well-defined
    error message.

2.) The scenario can be enabled by setting library to 6 compatibility mode in the reader and
    writer settings, i.e. settings.LibraryCompatibility = ODataLibraryCompatibility.Version6.
    In this case, the value is read and written as: "@odata.null":true

3.) The reader will support reading "@odata.null":true. However, it does not use
    the library compatibility mode to do so. The library compatibility mode
    has been added to the reader for the completness of the API.

AlanWong-MS committed:
Address all StyleCop warnings except for SA0102: syntax error. (#1092)

- Address all StyleCop warnings except for SA0102: syntax error. The SA0102 warning is vague and cannot be suppressed.

xuzhg committed:
Modify to pass the codegen t4 test cases

xuzhg committed:
add a test case for security schemes terms in authorization

xuzhg committed:
Fix the authorization xml error

biaol-odata committed:
Fixed rogue string detected by PoliCheck.

kosinsky committed:
Ensure that WebAPI aren't broken

kosinsky committed:
Rename PropertyAggregateExpression back to AggregateExpression

kosinsky committed:
Keep old aggregate token names

kosinsky committed:
Merge with master

mikepizzo committed:
Fix for not throwing and exception for an invalid enum value (#1112)

mikepizzo committed:
Fix support for containment paths in nav prop bindings (#1109)

- Fix numerous issues with containment nav prop bindings
- Fix annotation validation and contextUrl validation tests
- Fix issues with locating the proper contained entity set in FindNavigationTarget
- Fix nav prop binding issues with nav props on complex types
- Fix issues with reading/writing nav prop bindings for contained entity sets
- Add support for writing context for contained nav props under LibraryCompatibility.Version6
- Add support for entity types used exclusively in singletons and single-valued nav props without defining a key

kosinsky committed:
Merge remote-tracking branch 'full/master' into entity_set_aggregation

kosinsky committed:
Test to prove that entityset aggregations and functions without prefix can work together

mikepizzo committed:
Fix path for EdmContainedEntitySet (#1118)

kosinsky committed:
Ensure that we not using exception for control flow

kosinsky committed:
Merge remote-tracking branch 'full/master' into entity_set_aggregation

mikepizzo committed:
Enable writing type annotation for collections in full metadata (#1087)

* Enable writing type annotation for collections in full metadata; compatibility with ODL 6.x

* Put writing extra odat.type annotation under ODataLibraryCompatibility.Version6 setting.

* Revert "Put writing extra odata.type annotation under ODataLibraryCompatibility.Version6 setting."

This reverts commit 77a9e4cad97107b5d3ccebbab96179607c20a1b1.

kosinsky committed:
Extra tests for ApplyBinder

xuzhg committed:
Add StringComparison in String.Equal() method call

kosinsky committed:
Merge remote-tracking branch 'full/master' into entity_set_aggregation

kosinsky committed:
Cleaning up code

mikepizzo committed:
Make reading/writing navigationPropertyMappings dictionary threadsafe (#1133)

* Fix path for EdmContainedEntitySet

* Use ConcurrentDictionary for navigationPropertyMappings.
Also, don't read/write nav prop bindings that end in containment navigation properties.

voronov-maxim committed:
Fix metadata properties for apply ComputeTransformationNode (#1142)

- Fix write document metadata properties for apply ComputeTransformationNode
- Add corresponding test case

voronov-maxim committed:
Fix build uri from groupby navigation property with many child structural properties (#1143)

- Fix build uri from groupby navigation property with many child structural properties
- Add corresponding test case

xuzhg committed:
remove the Document element from Edm lib

dozer75 committed:
Added check if the MessageReaderSettings.ThrowIfTypeConflictsWithMetadata (#1158)

- Added check if the MessageReaderSettings.ThrowIfTypeConflictsWithMetadata is set to ensure that any metadata related checks are done if the flag is set, otherwise ignored.
- Fixed test so that the expected result corresponds to the settings of the behaviorkind that is currently running.

kosinsky committed:
Merge with master

kosinsky committed:
Minnor fixes (spaces, assert etc.)

kosinsky committed:
Rename files to reflect class names.

kosinsky committed:
Fix documentation tags

kosinsky committed:
Avoiud breaking changes and mark old properties obsolete

kosinsky committed:
Support Entity Set aggregation into ApplyClauseToStringBuilder

kosinsky committed:
More comment updates
